vwifi.c

```diff
@@ -403,15 +429,155 @@ static struct cfg80211_ops owl_cfg_ops = {
     .get_station = owl_get_station,
 };
 
+static int owl_ndo_open(struct net_device *dev)
+{
+    /* The first byte is '\0' to avoid being a multicast
+     * address (the first byte of multicast addrs is odd).
+     */
+    struct owl_context *owl = ndev_get_owl_context(dev)->owl;
+    char intf_name[ETH_ALEN] = {0};
+    snprintf(&intf_name[1], ETH_ALEN - 1, "%s", dev->name);
+    memcpy(dev->dev_addr, intf_name, ETH_ALEN);
+
+    INIT_LIST_HEAD(&owl->rx_queue);
+    netif_start_queue(dev);
+    return 0;
+}
+
+static int owl_ndo_stop(struct net_device *dev)
+{
+    struct owl_context *owl = ndev_get_owl_context(dev)->owl;
+    struct owl_packet *pkt, *is = NULL;
+    list_for_each_entry_safe (pkt, is, &owl->rx_queue, list) {
+        list_del(&pkt->list);
+        kfree(pkt);
+    }
+    netif_stop_queue(dev);
+    return 0;
+}
+
+static struct net_device_stats *owl_ndo_get_stats(struct net_device *dev)
+{
+    struct owl_context *owl = ndev_get_owl_context(dev)->owl;
+    return &owl->stats;
+}
+
+/*
+ * Receive a packet: retrieve, encapsulate and pass over to upper levels
+ */
+static void owl_rx(struct net_device *dev)
+{
+    struct owl_context *owl = ndev_get_owl_context(dev)->owl;
+    struct sk_buff *skb;
+    int hit_ours_pkt = 0;
+    int offset = 0;
+
+    struct owl_packet *pkt;
+    if (list_empty(&owl->rx_queue)) {
+        printk(KERN_NOTICE "owl rx: No packet in rx_queue\n");
+        goto out;
+    }
+    pkt = list_first_entry(&owl->rx_queue, struct owl_packet, list);
+    /* TODO: Modify the packet and forward to destination */
+    struct ethhdr *eth = (struct ethhdr *) &pkt->data[offset];
+    offset += sizeof(struct ethhdr);
+    /* Process ARP Request */
```

Comment:
I intend to merge this pull request without major refinement, which decouples protocol stuff from `owl_rx`. However, you shall add informative `FIXME` or `TODO` comments for further involvement.